### ISSUE TITLE
english and french text together causes crash

### ISSUE THREAD

**Issue by wrapperband** [karpathy/char-rnn#116]
I've been getting the error below with one particular (1.2MB) text file. 

I have traced it back (by halving the file) to some French text that was in the file. (i.e it's working as soon as I deleted the French text) 
I'm using UTF-8 (which I thought might be the problem) but that should allocate a number for each character? so should work best for languages (with other characters)...

stack traceback:
	[C]: in function 'index'
	./util/OneHot.lua:18: in function 'func'
	/home/tony/torch/install/share/lua/5.1/nngraph/gmodule.lua:289: in function 'neteval'
	/home/tony/torch/install/share/lua/5.1/nngraph/gmodule.lua:324: in function 'forward'
	train.lua:262: in function 'opfunc'
	/home/tony/torch/install/share/lua/5.1/optim/rmsprop.lua:32: in function 'rmsprop'
	train.lua:306: in main chunk
	[C]: in function 'dofile'
	...tony/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
	[C]: at 0x00405ea0

**Comment by hughperkins** [comment]
Are you sure you're not somehow generating negative indexes, eg by storing a number larger than 127 in a signed byte?

**Comment by wrapperband** [comment]
I'm pretty sure I wouldn't even know how to do that? 
I tested this sort of problem by removing text from a large training data that failed, untill I found the offending text, and the training data didn't fail. I then deleted all French. 

One point might be worth noting is that, it had previously worked with the same text in a smaller traing file, which is why I thought it was a format error. i.e sometimes 2 characters are interpreted as one larger / illegal number ....

I'm pretty disabled, so, I have to limit to what I can do and I was trying to optimise the size and length of historical data (rnn) is considering, a big job in itself, and one that could be "self aligned" to the data.
 
As this (problem) was apparent in the way it was learning my "training data" / use case. With my case, despite largest memory, most levels and big data, and un-like the folk format example with "abc", with the current char-rnn software version, it seemed to rapidly "unlearn" the long term structure. In the end, it could not learn back longer than a line. i.e bug fixing / testing not ML 

As I seem to be suffering from a number of "known issues" ,  I still have all my data, ready to retry on future versions, or fixes ..

**Comment by hughperkins** [comment]
You probably want to sprinkle `print` statements liberally in the code, to find out more about what is happening.  For example, what is the vocabulary size? ... oh, after checking ,seems it's printed already https://github.com/karpathy/char-rnn/blob/master/train.lua#L114  What vocabulary size are you seeing?

**Comment by wrapperband** [comment]
I can look more intoit after the holidays. but for example

input text 2.9MB, data 2.9MB vocal 2.4kb

**Comment by hughperkins** [comment]
basically, if vocabulary size is more than 255, then the input values are probably wrapping around.  If they're being stored in a signed byte, they will wrap around if the vocabulary size is more than 127.  The file sizes can probably be used to derive the vocabulary size, but better would be to find out what the program prints when you run it.

**Comment by wrapperband** [comment]
Cheers for that. I've moved to the new rig, with the R9 280, which has been flaky. I'll copy over a checkpoint from the old PC, and run it. 
Does it echo the vocab or do I need >> debug.txt or  -D?

pcs starting now so may anwser my own question

**Comment by hughperkins** [comment]
vocabulary 114 should be fine.  As long as it doesnt go above 127, you're almost certainly ok from vocab-size point of view.  Actually, even up to 255 *might* be fine, but someone would need to check more carefully, ie is the label stored in signed or unsigned byte?

> I'll copy over a checkpoint from the old PC, and run it. 

Note that I'd be careful about copying checkpoints from one machine to the other.  And especially between different GPUs and different CPU arcihtectures, since the endianness might be different.  Not sure which situations will definitely cause problems, but might be easier just to recreate new checkpoints, so as not to have to think about this.

> The character vocabulary in this dataset and the one in the saved checkpoint are not the same.

Recreating the checkpoint would also solve this issue.

> The new PC / AMD drivers and char-rnn were more flaky, i couldn't start a 700 4 layer on the new pc with the R9 280, or smaller ones actually.

Hmmmm, error messages?
